HIP/hip-351.md

```diff
@@ -69,18 +69,46 @@ When generating random numbers in a given range, if that range is not a power of
 
-It appears to be even more difficult for a user submitting a `RandomGenerateTransaction` to predict what the number will be. Because there is no way to predict what other transactions will come before it in consensus order, what their timestamps will be, and what their records will show as the result of executing them. So predicting the random number would appear to be difficult. But cannot be mathematically ruled out.
+It appears to be even more difficult for a user submitting a `UtilPrngTransaction` to predict what the number will be. Because there is no way to predict what other transactions will come before it in consensus order, what their timestamps will be, and what their records will show as the result of executing them. So predicting the random number would appear to be difficult. But cannot be mathematically ruled out.
 
 It is also theoretically possible that a flaw in SHA-2 will be found such that every SHA-384 hash has some property that is predictable in advance. For example, there might be some function of its bits that always returns a 0.  No such patterns are currently known, and there is currently no reason to expect such patterns to be discovered. But it cannot be mathematically ruled out.
 
-## How to Teach This
-
 ## Reference Implementation
 
+The solidity precompiled contract is to reside at address `0x169` and implement the below interface,
+
+```
+interface IPrngSystemContract {
+    // Generates a 256-bit pseudorandom seed using the first 256-bits of running hash of n-3 transaction record.
+    // Users can generate a pseudorandom number in a specified range using the seed by (integer value of seed % range)
+    function getPseudorandomSeed() external returns (bytes32);
+}
+```
+
+Smart contracts can use a new system contract to generate pseudorandom numbers. A typical use case that implements
+this interface might be,
+
+```
+import "./IPrngSystemContract.sol";
+
+contract PrngSystemContract {
+    address constant PRECOMPILE_ADDRESS = address(0x169);
+
+    function getPseudorandomSeed() external returns (bytes32 randomBytes) {
+        (bool success, bytes memory result) = PRECOMPILE_ADDRESS.call(
+            abi.encodeWithSelector(IPrngSystemContract.getPseudorandomSeed.selector));
+        require(success);
+        randomBytes = abi.decode(result, (bytes32));
```

Comment:
We might also include an example like,
```
/**
 * Returns a psuedorandom number in the range [lo, hi) 
 */
function getPseudorandomNumber(uint32 lo, uint32 hi) external returns (uint32) {
    (bool success, bytes memory result) = PRECOMPILE_ADDRESS.call(
        abi.encodeWithSelector(IPrngSystemContract.getPseudorandomSeed.selector));
    require(success);
    uint32 choice;
    assembly {
        choice := mload(add(result, 0x20))
    }
    return min + (choice % (max - min));
}        
```

Reply:
Thanks @tinker-michaelj  ! Added example.